From ad8f3bac5ee32951668f6da0530b00053ab631d1 Mon Sep 17 00:00:00 2001
From: Christian Ehrhardt <christian.ehrhardt@codasip.com>
Date: Fri, 21 Feb 2025 14:06:44 +0100
Subject: hugemmap: Workaround for LLVM miscompilation

LLVMs constant folding pass assumes that after you
compare a pointer with a constant or otherwise a known
value it know the value of the pointer. However, this is
not true for CHERI wher the comparison only gives you the
address part of th pointer.

Add volatiles to some pointer variables to prevent this
optimization.
---
 testcases/kernel/mem/hugetlb/hugemmap/hugemmap13.c | 2 +-
 testcases/kernel/mem/hugetlb/hugemmap/hugemmap14.c | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

Upstream-Status: Inappropriate [ CHERI Specific ]

diff --git a/testcases/kernel/mem/hugetlb/hugemmap/hugemmap13.c b/testcases/kernel/mem/hugetlb/hugemmap/hugemmap13.c
index f8c36640d..256be23a9 100644
--- a/testcases/kernel/mem/hugetlb/hugemmap/hugemmap13.c
+++ b/testcases/kernel/mem/hugetlb/hugemmap/hugemmap13.c
@@ -32,7 +32,7 @@ static int page_size;
 
 static void run_test(void)
 {
-	void *p, *q = NULL;
+	void * volatile p, * volatile q = NULL;
 	unsigned long long lowaddr;
 	unsigned long long below_start;
 	unsigned long long above_end;
diff --git a/testcases/kernel/mem/hugetlb/hugemmap/hugemmap14.c b/testcases/kernel/mem/hugetlb/hugemmap/hugemmap14.c
index c54a746b3..0a467f69a 100644
--- a/testcases/kernel/mem/hugetlb/hugemmap/hugemmap14.c
+++ b/testcases/kernel/mem/hugetlb/hugemmap/hugemmap14.c
@@ -32,7 +32,7 @@ static int page_size;
 
 static void run_test(void)
 {
-	void *p, *q = NULL, *r = NULL;
+	void * volatile p, * volatile q = NULL, * volatile r = NULL;
 	unsigned long long lowaddr, highaddr;
 	unsigned long long below_start;
 	unsigned long long above_end;
-- 
2.43.5

